scanner/api/source.h

```diff
@@ -0,0 +1,135 @@
+/* Copyright 2018 Carnegie Mellon University
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "scanner/api/kernel.h"
+#include "scanner/api/enumerator.h"
+#include "scanner/util/common.h"
+#include "scanner/util/profiler.h"
+
+#include <vector>
+
+namespace scanner {
+
+//! Parameters provided at instantiation of Source node
+struct SourceConfig {
+  std::vector<std::string> output_columns;
+  std::vector<proto::ColumnType> output_column_types;
+  std::vector<u8> args;  //! Byte-string of proto args if given.
+  i32 node_id;
+};
+
+/**
+ * @brief Interface for reading data in a computation graph.
+ *
+ * Sources are how Scanner's computation graphs read data from outside the
+ * system.
+ */
+class Source {
+ public:
+  Source(const SourceConfig& config) {}
+
+  virtual ~Source(){};
+
+  /**
+   * @brief Checks if Source arguments are valid.
+   *
+   * Only useful if your Source has its own custom Protobuf arguments.
+   */
+  virtual void validate(proto::Result* result) { result->set_success(true); }
+
```

Comment:
Why does Sink have `new_stream` but Source does not?

Reply:
The reason Sink has `new_stream` is to pass the `per-stream` arguments that are specified when you bind data to the Op during Job creation. Source doesn't have an equivalent because it already receives per-element arguments for each element in `read`. Can you think of a need for a `per-stream` argument rate for Source?